django/db/models/sql/query.py

```diff
@@ -1299,13 +1299,36 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
             self.check_filterable(value)
 
         clause = self.where_class()
+        alias = self.get_initial_alias()
         if reffed_expression:
             condition = self.build_lookup(lookups, reffed_expression, value)
             clause.add(condition, AND)
-            return clause, []
+            targets = reffed_expression.get_source_expressions()
+            if targets and isinstance(targets[0], Col):
+                target = targets[0].target
+                lookup_type = condition.lookup_name
+                clause.add(condition, AND)
+                if (
+                    current_negated and
+                    lookup_type != 'isnull' and
+                    condition.rhs is not None and
+                    self.is_nullable(target)
+                ):
+                    # The condition added here will be SQL like this:
+                    # NOT (col IS NOT NULL), where the first NOT is added in
+                    # upper layers of code. The reason for addition is that if col
+                    # is null, then col != someval will result in SQL "unknown"
+                    # which isn't the same as in Python. The Python None handling
+                    # is wanted, and it can be gotten by
+                    # (col IS NULL OR col != someval)
+                    #   <=>
+                    # NOT (col IS NOT NULL AND col = someval).
+                    lookup_class = target.get_lookup('isnull')
+                    col = self._get_col(target, target, alias)
+                    clause.add(lookup_class(col, False), AND)
+            return clause, ()
```

Comment:
We can try to refactor [existing code](https://github.com/django/django/blob/a9cf954e6174450057ea1065aa2ccbbd12f59b65/django/db/models/sql/query.py#L1351-L1375) and add some internal hook instead of having similar (identical?) logic in two places.

Reply:
Also, I don't think it's a complete solution as annotations on the LHS still don't work properly, see failing test:
```python
    def test_exclude_nullable_fields_annotation(self):
        from django.db.models.functions import Abs
        number = Number.objects.create(num=1, other_num=1)
        Number.objects.create(num=2, other_num=2, another_num=2)
        self.assertSequenceEqual(
            Number.objects.annotate(x=Abs('another_num')).exclude(other_num=F('x')),
            [number],
        )   
        self.assertSequenceEqual(
            Number.objects.annotate(x=Abs('another_num')).exclude(num=F('x')),
            [number],
        )   
```

Reply:
@felixxm in my view, below test code is our expected result.
```
    def test_exclude_nullable_fields_annotation(self):
        from django.db.models.functions import Abs
        number = Number.objects.create(num=1, other_num=1)
        Number.objects.create(num=2, other_num=2, another_num=2)
        self.assertSequenceEqual(
            Number.objects.annotate(x=Abs('another_num')).exclude(other_num=F('x')),
            [],
        )
        self.assertSequenceEqual(
            Number.objects.annotate(x=Abs('another_num')).exclude(num=F('x')),
            [],
        )
```
Because I expect that query makes like below SQL. I think another_sum is nullable, so need to add is not null.
it means return 0 rows.

```
-- First Queryset
SELECT
	num,
	other_num,
	another_num
FROM
	number
WHERE
	NOT(another_num IS NOT NULL AND other_num = abs(another_sum))
	
-- Second Queryset
SELECT
	num,
	other_num,
	another_num
FROM
	number
WHERE
	NOT(another_num IS NOT NULL AND num = abs(another_sum))	

```

Let me know if I'm misunderstanding.

Reply:
@felixxm At the first, I wanna reuse  [existing code](https://github.com/django/django/blob/a9cf954e6174450057ea1065aa2ccbbd12f59b65/django/db/models/sql/query.py#L1351-L1375) but I can't. Because for reusing that code, we have to use some code such as self.setup_joins, self.trim_joins. but setup_joins is not designed for reffed_expression.(just my opinion, as you know that this is my second fix, so I may not understand the code because I am not familiar with it yet.)

Reply:
If you have some tip about this, please give me. i also wanna improve this.

Reply:
> @felixxm in my view, below test code is our expected result. 

No, it's not. `1 != NULL` so why it's expected that `number` is excluded? see 512da9d5855 and ticket-23797 for more details.

Reply:
@felixxm You're right. but it seems to fail in the 'master' branch, too. and that query is not used changed code.

Reply:
my changs affect when LHS is annotation field and RHS is not field.

Reply:
> my changs affect when LHS is annotation field and RHS is not field.

I know, but this is also an issue with nullable annotation so we should fix it in this PR. If it is a different branch in `build_filter()` then we have another reason to add some hook.

Reply:
@felixxm Ok. it makes sense. I will check this also.